python/spec/dependabot/python/file_parser_spec.rb

```diff
@@ -158,7 +158,10 @@
 
         it "has the right details" do
           expect(dependency).to be_a(Dependabot::Dependency)
-          expect(dependency.name).to eq("psycopg2")
+          expect(dependency.name).to start_with("psycopg2[")
+          expect(dependency.name).to end_with("]")
+          expect(dependency.name).to include("foo")
+          expect(dependency.name).to include("bar")
```

Comment:
Would be clearer to expect the full name with one exepect:

```suggestion
          expect(dependency.name).to eq("psycopg2[foo,bar]")
```

Reply:
I agree! However, I found during testing that the order of 'foo' and 'bar' can change from run to run. This doesn't affect the functionality of the program at all, but it means that the obvious test becomes flaky.
I see two options: this, or sorting the 'extras' list that comes back from the python script alphabetically. I felt it was probably less confusing to write the test like this, but happy to change if that's what you'd prefer.

Reply:
Ah I see, thinking it would be good to make this deterministic somehow otherwise you could get a flapping diff between Dependabot update runs which would be quite confusing. Do you know if it's possible to get the original extras string so we don't need to pass around the array and join?

Reply:
So the way it works is this:
When parsing the files, extras are added to a python set. Because this is not hashable, I cast it to a list, which is returned as a JSON array and then ultimately interpreted as a Ruby array. It is _probably_ this change from a set to a list that picks one order over another, but it should only happen this once. That is, as soon as it gets in to the Ruby code, I believe the array order is fixed and so the casting to a string is also fixed.

So we're pretty much blocked by the setup.py parsing code when it comes to getting the original string. Since that ultimately delegates to a method in the pip library itself, it'd be unwise to start messing with that too much. 

However, I will take a look in that area of the code and do some power googling, see if there's a way to enforce that the list casting respects the order that things were added to the set, just in case there's something I've overlooked there.

Reply:
So I looked at the code again. There's essentially nothing we can do about the fact that extras are being inserted into a set, and that insertion does not 'remember' the order in which they are added. There will therefore always be a chance that dependabot will produce an 'extras' output that is different from the input.

The best we can probably do is then sorting this list, since then at least we have that determinism. I will therefore change this PR to reflect that.

Reply:
Nice one, that sounds good to me! We had a case in the past where npm would reshuffle a config key in the lockfile each time dependabot ran and it caused all kinds of confusion 🤠